ggml.c

```diff
@@ -15126,6 +15055,29 @@ static void ggml_compute_forward_cross_entropy_loss_f32(
 
 }
 
+static void ggml_compute_forward_cross_entropy_loss_finalizer(
+        struct ggml_compute_params * params,
+        struct ggml_tensor * dst) {
+
+    const struct ggml_tensor * src0 = dst->src0;
+
+    switch (src0->type) {
+        case GGML_TYPE_F32: {
+            GGML_ASSERT(params->ith == 0);
+
+            float * sums = (float *) params->wdata;
+            // TODO: handle transposed/permuted matrices
+            float * dp = (float *) dst->data;
+            ggml_vec_sum_f32(params->nth, dp, sums);
+            dp[0] *= -1.0f;
+        } break;
+    default:
+        {
+            GGML_ASSERT(false);
+        } break;
+    }
+}
+
```

Comment:
Oops, I have missed that we are now using `GGML_TASK_FINALIZE` here.

@xaedes 
I'm thinking about deprecating `GGML_TASK_FINALIZE` and this is the only exception that currently uses it. Do you think there is a way to make this work without the need for "finalize" pass?

Reply:
Hm... I don't think so. 
There needs to be some finalizing synchronization for computing aggregated values.
Similarly `ggml_compute_forward_sum_f32` (or mean) currently is not, but _could_ be parallelized. 
Then it also would need a final synchronization step.
The only way I see is to split it in multiple operations, storing intermediate tensors and then use a not-parallelizeable-by-design aggregating operation on the intermediate tensor.

I think finalizer lookup table is a good idea to avoid the finalize stage for operations that don't need it.

Reply:
Atomic (add) operations could also be used to avoid finalizing synchronization, but such atomic functions often only work on integral types.

Reply:
> Atomic (add) operations could also be used to avoid finalizing synchronization, but such atomic functions often only work on integral types.

Current "problem" can be resolved by any kind of lock, but that's ugly and low efficient. Using atomic directly in compute code is a bad smell, that tends to result in difficult situations.

Let me share a post about linux spinlock, not quite relevant to current topic https://www.realworldtech.com/forum/?threadid=189711&curpostid=189723v

Quote:
```
... There's a reason why you can find decades of academic papers on locking. Really. It's hard.
```

Recently I tried c++ locks, feels bad, really. Any kind of logically correct locking has to be implemented by the underlying platform, but cross-platform sucks, CPU matters.

Reply:
Yea - locks are not good.

@mqy 
Can you try to adapt this solution proposed by @zrm in a recent PR:

https://github.com/ggerganov/llama.cpp/pull/1556/commits/9d058c2096b9f1f300e1ee16f5740a6a0a342917

We ended up not using it, but I think in the current situation it will be the best option.

Reply:
Sure, the bool array is good, no longer have to split finalizers, I think @ xaedes will feel happy :)
I'll do that, add the `GGML_OP_HAS_INIT` array as well.

Reply:
Also deprecated GGML_TASK_INIT.
No longer touch runner codes, because that would modify or delete many lines,  tedious and error prone.
We'd cleanup them later step by step.

@ggerganov pls have a look.

Reply:
The macOS-make-latest build failed by "brew update failed"